File: src/leetcode/java/26.remove-duplicates-from-sorted-array.java
import java.util.*;

public class Solution {
    public int removeDuplicates(int[] nums) {
        if (nums.length == 0 || nums.length == 1) return nums.length;
        int i = 1, j = 1;
        int d = nums[0]; // duplicate
        while (i < nums.length && j < nums.length) {
            if (nums[i] <= nums[i - 1]) {
                if (j <= i) j = i + 1;
                if (nums[i] > d) d = nums[i];
                while (j < nums.length && nums[j] <= d) {
                    j++;
                }
                if (j < nums.length) {
                    d = nums[j];
                    swap(nums, i, j);
                } else {
                    break;
                }
            }
            i++;
        }

        return i;
    }
    private void swap(int[] nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
}

File: leetcode/javascript/26.remove-duplicates-from-sorted-array.js
/*
 * [26] Remove Duplicates from Sorted Array
 *
 * https://leetcode.com/problems/remove-duplicates-from-sorted-array
 *
 * algorithms
 * Easy (35.52%)
 * Total Accepted:    241.2K
 * Total Submissions: 679.3K
 * Testcase Example:  '[]'
 *
 * 
 * Given a sorted array, remove the duplicates in place such that each element
 * appear only once and return the new length.
 * 
 * 
 * Do not allocate extra space for another array, you must do this in place
 * with constant memory.
 * 
 * 
 * 
 * For example,
 * Given input array nums = [1,1,2],
 * 
 * 
 * Your function should return length = 2, with the first two elements of nums
 * being 1 and 2 respectively. It doesn't matter what you leave beyond the new
 * length.
 * 
 */
function swap(nums, i, j) {
  let t = nums[i];
  nums[i] = nums[j];
  nums[j] = t;
}
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
  if (nums.length === 0) return 0;
  let a = 0, b = 0, lastB = nums[b];
  while (true) {
    a++;
    b++;
    while (nums[b] === lastB && b < nums.length) {
      b++;
    }
    if (b >= nums.length) break;
    lastB = nums[b];
    swap(nums, a, b);
  }
  return a;
};
